homeassistant/components/ring/entity.py

```diff
@@ -29,25 +29,23 @@
 
 
 def exception_wrap[_RingBaseEntityT: RingBaseEntity[Any, Any], **_P, _R](
-    func: Callable[Concatenate[_RingBaseEntityT, _P], _R],
-) -> Callable[Concatenate[_RingBaseEntityT, _P], _R]:
+    async_func: Callable[Concatenate[_RingBaseEntityT, _P], Coroutine[Any, Any, _R]],
+) -> Callable[Concatenate[_RingBaseEntityT, _P], Coroutine[Any, Any, _R]]:
     """Define a wrapper to catch exceptions and raise HomeAssistant errors."""
 
-    def _wrap(self: _RingBaseEntityT, *args: _P.args, **kwargs: _P.kwargs) -> _R:
+    async def _wrap(self: _RingBaseEntityT, *args: _P.args, **kwargs: _P.kwargs) -> _R:
         try:
-            return func(self, *args, **kwargs)
+            return await async_func(self, *args, **kwargs)
         except AuthenticationError as err:
-            self.hass.loop.call_soon_threadsafe(
-                self.coordinator.config_entry.async_start_reauth, self.hass
-            )
+            self.coordinator.config_entry.async_start_reauth(self.hass)
```

Comment:
Can't we raise ConfigEntryAuthFailed?

Reply:
If we have tests for this I am curious if this works as well

Reply:
That only works when running within the data coordinator which catches the ConfigEntryAuthFailed exception and starts reauth. Outside of the coordinator we have to do it this way, n.b. we do the same in [tplink here](https://github.com/home-assistant/core/blob/b541d214ed8859a95cfce115610289f5675d0a4e/homeassistant/components/tplink/entity.py#L96)

As per request I made this quick update to test it out:
```py
        try:
            return await async_func(self, *args, **kwargs)
        except AuthenticationError as err:
            #self.coordinator.config_entry.async_start_reauth(self.hass)
            raise ConfigEntryAuthFailed(err) from err
            #raise HomeAssistantError(err) from err
```
and re-ran the test suite and the tests that raise Authentication errors and check that re-auth has been started fail:
```
FAILED tests/components/ring/test_camera.py::test_motion_detection_errors_when_turned_on[Authentication] - assert False == True
FAILED tests/components/ring/test_light.py::test_light_errors_when_turned_on[Authentication] - assert False == True
FAILED tests/components/ring/test_siren.py::test_siren_errors_when_turned_on[Authentication] - assert False == True
FAILED tests/components/ring/test_switch.py::test_switch_errors_when_turned_on[Authentication] - assert False == True

Results (22.38s):
      74 passed
       4 failed
         - tests/components/ring/test_camera.py:160 test_motion_detection_errors_when_turned_on[Authentication]
         - tests/components/ring/test_light.py:87 test_light_errors_when_turned_on[Authentication]
         - tests/components/ring/test_siren.py:126 test_siren_errors_when_turned_on[Authentication]
         - tests/components/ring/test_switch.py:94 test_switch_errors_when_turned_on[Authentication]
```
N.B. When I did this exception wrap for tplink which is already async I initially used `call_soon` which was agreed to be unnecessary.  Discussion [here for reference](https://github.com/home-assistant/core/pull/114919#issuecomment-2041443056)